### PR TITLE
GRK-143 added lat long converter via OpenStreetMap

### PR DESCRIPTION
**Pull request by eyalacato**

### CODE REVIEW

File: src/rest_api/class-openagenda-controller.php [hunk 1]
```diff
@@ -1439,6 +1439,16 @@ public function prepare_item_for_response( $item, $request ) {
 			$item_data[ $meta_key ] = $meta_value;
 		}
 
+		// Create Lat/long based on address data from OSM.
+		if ( ! empty( $item_data['location_address'] ) && ! empty( $item_data['location_zipcode'] ) && ! empty( $item_data['location_city'] ) ) {
+			$address     = $item_data['location_address'] . ', ' . $item_data['location_zipcode'] . ' ' . $item_data['location_city'];
+			$address     = str_replace( ' ', '+', $address );
```
Comment (rmpel) [inline src/rest_api/class-openagenda-controller.php:1445]: What is the reason for this replacement? I suspect URL encoding, but then, should php functions for that not be used instead?

File: src/rest_api/class-openagenda-controller.php [hunk 2]
```diff
@@ -1602,6 +1612,16 @@ public function prepare_location_for_response( $item, $request, $include_events
 			$item_data[ $meta_key ] = $meta_value;
 		}
 
+		// Create Lat/long based on address data from OSM.
+		if ( ! empty( $item_data['address'] ) && ! empty( $item_data['zipcode'] ) && ! empty( $item_data['city'] ) ) {
```
Comment (rmpel) [inline src/rest_api/class-openagenda-controller.php:1616]: Duplicated code, is there perhaps a DRY optimisation possible?

File: src/rest_api/class-openagenda-controller.php [hunk 3]
```diff
@@ -1736,4 +1750,44 @@ function ( $a, $b ) {
...
+		}
+
+		$address_full = $address . ', ' . $zipcode . ' ' . $city;
+		$address_full = rawurlencode( $address_full );
+
+		// Get the address data from OSM (OpenStreetMap).
+		$osm_url     = 'https://nominatim.openstreetmap.org/search?q=' . $address_full . '&format=json&addressdetails=1';
+		$osm_address = wp_remote_get( $osm_url );
+
+		if ( ! $osm_address ) {
```
Comment (rmpel) [inline src/rest_api/class-openagenda-controller.php:1775]: what if osm_address is a WP_Error?

File: src/rest_api/class-openagenda-controller.php [hunk 4]
```diff
@@ -1736,4 +1750,44 @@ function ( $a, $b ) {
...
+			return null;
+		}
+
+		$osm_address = json_decode( $osm_address['body'] );
```
Comment (rmpel) [inline src/rest_api/class-openagenda-controller.php:1779]: what if decode fails?

File: src/rest_api/class-openagenda-controller.php [hunk 5]
```diff
@@ -1736,4 +1750,44 @@ function ( $a, $b ) {
...
+
+		if ( ! $osm_address[0]->lat || ! $osm_address[0]->lon ) {
+			return null;
```
Comment (rmpel) [inline src/rest_api/class-openagenda-controller.php:1782]: osm_address could be something other than an array, with one element, that is an object, with a lat
please consider extra checks

File: src/rest_api/class-openagenda-controller.php (hunk 3 above)
Comment (eyalacato) [reply src/rest_api/class-openagenda-controller.php]: I added an extra check

File: src/rest_api/class-openagenda-controller.php (hunk 4 above)
Comment (eyalacato) [reply src/rest_api/class-openagenda-controller.php:1780]: I added an extra check

File: src/rest_api/class-openagenda-controller.php (hunk 5 above)
Comment (eyalacato) [reply src/rest_api/class-openagenda-controller.php:1783]: I added an extra check

**Review by rmpel — APPROVED** [review]

**Pull request merged**